gobblin-core/src/test/java/gobblin/converter/EnvelopeSchemaConverterTest.java

```diff
@@ -0,0 +1,96 @@
+package gobblin.converter;
+
+import com.google.common.base.Optional;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import gobblin.configuration.ConfigurationKeys;
+import gobblin.configuration.WorkUnitState;
+import gobblin.converter.EnvelopeSchemaConverter;
+import gobblin.converter.filter.AvroProjectionConverter;
+import gobblin.converter.filter.AvroSchemaFieldRemover;
+import gobblin.metrics.kafka.KafkaAvroSchemaRegistry;
+import java.util.Properties;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DecoderFactory;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+public class EnvelopeSchemaConverterTest {
+
+  private final String SCHEMA_KEY = "testKey";
+
+  private GenericRecord mockInputRecord = mock(GenericRecord.class);
+  private GenericRecord mockOutputRecord = mock(GenericRecord.class);
+  private Schema mockSchema = mock(Schema.class);
+
+  class KafkaAvroSchemaRegistryForTest extends KafkaAvroSchemaRegistry {
+    public KafkaAvroSchemaRegistryForTest(Properties props) {
+      super(props);
+    }
+
+    @Override
+    public Schema getSchemaByKey(String key) {
+      if (key.equals(SCHEMA_KEY)) {
+        return mockSchema;
+      } else {
+        return null;
+      }
+    }
+  }
+
+  class EnvelopeSchemaConverterForTest extends EnvelopeSchemaConverter {
+    @Override
+    public EnvelopeSchemaConverterForTest init(WorkUnitState workUnit) {
+      if (workUnit.contains(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY)) {
+        String removeFieldsPropName = workUnit.getProp(ConfigurationKeys.EXTRACT_TABLE_NAME_KEY) + AvroProjectionConverter.REMOVE_FIELDS;
+        if (workUnit.contains(removeFieldsPropName)) {
+          this.fieldRemover = Optional.of(new AvroSchemaFieldRemover(workUnit.getProp(removeFieldsPropName)));
+        } else {
+          this.fieldRemover = Optional.absent();
+        }
+      }
+      this.registry = new KafkaAvroSchemaRegistryForTest(workUnit.getProperties());
```

Comment:
If I am not mistaken the only different between EnvelopeSchemaConverter and EnvelopeSchemaConverterForTest is the Kafka schema registry implementation. If that is the case, I suggest that you use dependency injection instead of having to copy/paste the code which is error prone and the copies will go out-of-sync next time we change EnvelopeSchemaConverter.init() .


You can define:
public interface gobblin.metrics.kafka.KafkaSchemaRegistryFactory<K, S> {
   KafkaSchemaRegistry<K, S> create(Properties config);
}

Then have two implementations:
- KafkaAvroSchemaRegistryFactory or KafkaAvroSchemaRegistry.Factory
- KafkaAvroSchemaRegistryForTestFactory or KafkaAvroSchemaRegistryForTest.Factory

You can have a new configuration property for EnvelopeSchemaConverter that will contain the factory class. The default value should be KafkaAvroSchemaRegistryFactory/KafkaAvroSchemaRegistry$Factory but you can override to KafkaAvroSchemaRegistryForTestFactory/KafkaAvroSchemaRegistryForTest$Factory for testing purposes.

Then your init method can lookup the above factory and instantiate it dynamically.